### ISSUE TITLE
Support for 'shived' dependencies to ease deployment

### ISSUE THREAD

**Issue by sdementen** [xlwings/xlwings#1821]
shiv (https://github.com/linkedin/shiv) allows to pack all dependencies (site-packages/...) from a project in a single auto-executable file (.pyz using shebang to start it), including non pure python dependencies (compiled code, etc).

It would be nice to have the ability for xlwings to use such a zip package that could be deployed alongside an excel workbook.

**Comment by fzumstein** [comment]
This still does require python to be installed on the target machine though?

**Comment by sdementen** [comment]
Yes, just the base interpreter. This can be easily automated in a entreprise environnement.
I have a WIP branch on my fork of xlwings for this shiv integration.
I am just blocked by the shebang line that cannot have variable (like ~ or %HOME%) making it tricky to host the shiv file (a .pyz file) into a shared folder if the base interpreter is installed within the HOME folder (if install on a fixed path, there is no issue and it works very well)

**Comment by FabianGoessling** [comment]
Hi sdementen, what's the state of your implementation? I'm currently testing different setups for using a .pyz file and your changes in sdementen:shiv-support solved multiple problems, e.g. copying the win32 api dlls and the xlwings.dll loading in the .xlam.
Regarding the shebang line: We can add the .pyz file path into the pyxl call (don't remember the exact function name in the xlam addin), e.g. by a "PYZ PATH" variable in the .conf file. Then the interpreter can be set to a real path (not using the shebang at all) and there is no problem with using the pyz on a shared folder.

@fzumstein I guess the shiv support is not anywhere on your roadmap, but what about adding two more optional settings for a) adding something to the interpreter call (including interpreter options as e.g. -S) and b) loading the dll from a custom path ( see #1803) ?

**Comment by sdementen** [comment]
Hi @FabianGoessling, I have made some small changes to the xlwings VBA code + some preamble code in my shiv package to have a smooth integration.
Essentially:
- in the pyz file, I have added a preamble that:
  1. ensure that all libraries in "site_packages/pywin32_system32/*" are copied to the "site_packages/win32/lib" folder
  2. ensure that all "site_packages/xlwings??-*.dll" are copied to "site_packages/xlwings??.dll" (ie remove the version as otherwise an hassle when udpating xlwings
  3. check the sys.argv argument for the "--py" command and if so will exit immediately after having printed on the console the path to the parent of "site_packages" ==> will be called later in excel/xlwings to get the path to the unzipped site_packages and be able to find the xlwings??.dll)
- in xlwings:
  1. I have added a function that call the pyz file with the --py argument, catch and return the console output, giving me the path to the unzipped site_packages (with some caching logic)
```vba
Public shiv_path As String
Public shiv_key As String
Function GetShivPath(interpreter) As String
    ' run the interpreter and return the fist line of the stdout
    Dim key As String
    If Right(interpreter, 4) <> ".exe" Then
        key = interpreter & FileDateTime(interpreter)
        If key <> shiv_key Or shiv_path = "" Then
            cmd = "cmd.exe /C """ + interpreter & """ --py"
            shiv_path = CreateObject("WScript.Shell").Exec(cmd).stdout.readline
            shiv_key = key
        End If
    Else
        shiv_path = ""
        shiv_key = ""
    End If
    GetShivPath = shiv_path
End Function
```
  2. I have added this path to the paths xlwings use to find the DLL
```vba
     If (PYTHON_WIN <> "python" And PYTHON_WIN <> "pythonw") Or (CondaPath <> "" And CondaEnv <> "") Then
        If LoadLibrary(ParentFolder(PYTHON_WIN) + "\" + XLPyDLLName) = 0 Then  ' Standard installation
            If LoadLibrary(ParentFolder(ParentFolder(PYTHON_WIN)) + "\" + XLPyDLLName) = 0 Then  ' Virtualenv
                If LoadLibrary(GetShivPath(PYTHON_WIN) + "\site-packages\" + XLPyDLLName) = 0 Then  ' Shiv
                    Err.Raise 1, Description:= _
                    "Could not load " + XLPyDLLName + " from either of the following folders: " _
                    + vbCrLf + ParentFolder(PYTHON_WIN) _
                    + vbCrLf + ", " + ParentFolder(ParentFolder(PYTHON_WIN))
                    End If
            End If
        End If
    End If
```
  3. I set as the python interpreter the .pyz file

and voilà, a fully relocatable xlwings powered workbook!

I deploy on a sharepoint/onedrive folder:
  1. a python distribution (shared by all tools, the "embedded" version that is quite compact and sync quickly via onedrive)
  2. a folder per tool/workbook with:
    a) the .pyz file which contains only the zipped virtual env, named as ".cmd" to make it executable without needing the registry configuration about .pyz files being executable (as this would require the user to install python on their laptop. The shebang line points to the shared python distribution of (1). 
    b) a src.zip with the workbook specific python code, that is added also via the shiv preamble to the PYTHON_PATH.
  I split the venv and the src as the venv rarely changes but may be huge (if you include pandas/scipy) and the src is usually quite small and may be updated frequently (specially during the "test" phase)
    c) the workbook
    (these 3 files are deployed via a script that a) regenerates the .pyz if it detects a change in the venv configuration, b) compresses the source code in the .zip, c) copies the workbook to the target folder and adapts the xlwings.conf sheet

I use also most of this logic (except xlwings specifics) to deploy dash/plotly, jupyter, scripts, etc to users which haven't python installed on their machine.

Regarding xlwings, I think indeed the shiv integration goes well beyond the project (quite a burden for @fzumstein) and can be developed/packaged independently.
The support that would be needed within xlwings would be:
- ability to specify a command in the xlwings.conf to retrieve the xlwings.dll fill path (with some caching)
- remove the hard coding of the dll version in the VBA. This could be worked around if for the previous point, the dll file names with version would be sent to the command as then the command could fake the version,i.e the command would get "xlwings64-0.27.7.dll" in the arguments and could rename on the fly the latest version, eg xlwings64-0.29.1.dll to the expected version...

**Comment by fzumstein** [comment]
Hey @FabianGoessling  and @sdementen! Let's start with the good news:

* Refactoring the dll handling is a relatively high priority for me: https://github.com/xlwings/xlwings/issues/2026 Maybe I can get some new ideas from this PR
* Being able to specify the Python command is also somewhere on the roadmap albeit with less pressure than the previous point and for other reasons (autostarting xlwings Server on localhost)

Other than that, I don't really understand the point of shiv I guess? Why can't you just zip/unzip a full virtual env (or even full Python installation) on the target machine? I know users who do that with conda envs, and it seems to works nicely without any workarounds/hacks/[performance penalties](https://github.com/linkedin/shiv/issues/208).

xlwings supports local Python installations and freezers, and in the PRO version 1-click installers and xlwings Server. Shiv isn't mainstream enough for me consider it to include at this point...

**Comment by FabianGoessling** [comment]
Hi, 

@fzumstein 
Sounds good, looking forward to the changes, always good to have less of my own code to look after.

For me the shiv thing is the alternative to a pex file, which serves as a build artifact of e.g. the [pants](https://www.pantsbuild.org) build tool. As I won't be able to make pex run in our windows environment soon, shiv achieves the same, it yields a reproducible build artifact. Under the hood this is nothing more then a python [zip app](https://docs.python.org/3/library/zipapp.html), but with all 3-rd party deps included. As such this is essentially the same as the zip/unzip variant but with handy tools, such as build-ids, automatic unzipping, custom entry points, and others. Invoking is stupidly easy, no need to write a script to unzip the env and so on. If i recall correctly pyinstaller worked the same, it zips everything and then unzips for running the (pseudo-).exe, but for me the setup was way more complicated than the shiv setup.

Moreover due to my pseudo-monorepo setup  with multiple packages/libs and a .venv/poetry combination with editable installs, I can't just zip my .venv, but would require building a versionated fresh .venv nevertheless for zipping a distribution.
Performance wise I cannot confirm that issue, my setup is a server process, which keeps running and works with **all** workbooks. So once the process runs, I won't shut it down anytime soon. I think i did some hacking there to, something with the calling /active book in the xlam, don't remember atm. That might be another nice configuration possibility: *Do open new python process or keep a single one.*

Looking forward to @sdementen reasons to use shiv.

@sdementen 
Thanks for your reply, appreciate it, very helpful. I went the < "DLL PATH" variable in the .conf file > route and set the variable while first installing my addin. Thus no need to invoke another cmd call, but of course no "automatic" update of the DLL path, if I update the .pyz file at the moment. But as my setup uses a custom excel ribbon addin and accompanying .conf file (unique for all installations, so addin and .conf in a shared folder is not possible), an install is necessary anyway (by calling a cli incorporated in the .pyz). Still i like the idea of never thinking about the DLL path again, so I might adopt your extra cmd call.

**Comment by sdementen** [comment]
My use case is about zero-install deploymznt of Python apps (xlwings or scripts or jupyter environments or dash apps or ...). 
I use onedrive to ensure the distribution of both the python distribution (portable/embedded python) as well as the venv and the source code. 
When the source code of the tool is updated, only the zip file with the source code (and the excel workbook if it is an xlwings powered app) is updated by copying the file on the onedrive.
Any user that is syncing the applications folder (with the python dist, and one folder per app with the venv shived and the sex.zip file) can immediately use the apps without having to run any app installer or installing any python distribution => zero install apps. The whole applications folder must just be synced once in onedrive by the user so that the folder is replicated on the user laptop. Any update is then automagically synced without any user interaction.
Within this framework, I have xlwings powered workbook running smoothly, the user not even knowing python is behind it.
If not fully clear, I can elaborate more in detail .
Today, I have an internal python package that any python dev can use to deploy simply his/her app in this way (without knowing about shiv & co).
